### PR TITLE
feat: Cudagraphs integration for Torch-TRT + Non-default Stream Utilization

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- Add option to enable Cudagraphs in the runtime for additional acceleration of TRT engines
- Add C++ and Python toggles and well as full integration for C++ and Python runtimes
- Add support for dynamic shape cases via shape keys with cache invalidation
- Add test cases for cudagraphs support
- Add support for non-default CUDA Stream usage by default

Addresses #2736 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: core/runtime/execute_engine.cpp [hunk 1]
```diff
@@ -168,27 +210,66 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
       auto dims = core::util::toVec(out_shape);
       auto type = util::TRTDataTypeToScalarType(compiled_engine->exec_ctx->getEngine().getTensorDataType(name.c_str()));
       outputs[pyt_idx] = std::move(at::empty(dims, {at::kCUDA}).to(type).contiguous());
+
+      // In cudagraphs mode, the allocated output buffers are stored for reuse
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:214]: Is there a downside to always using the buffers?

Comment (narendasan) [reply core/runtime/execute_engine.cpp:214]: Also what is the required lifetime of these buffers, can they be deallocated at any point or do they need to be reserved as long as the cuda graph is active?

File: core/runtime/execute_engine.cpp [hunk 2]
```diff
@@ -168,27 +210,66 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
...
-          std::make_unique<torch::autograd::profiler::RecordProfile>(compiled_engine->enqueue_profile_path);
-    }
+  std::unique_ptr<torch::autograd::profiler::RecordProfile> enqueue_profiler_guard;
+
+  // nvinfer1::IExecutionContext::enqueue is not thread safe and we need a mutex for it.
+  std::unique_lock<std::mutex> lock(compiled_engine->mu);
+
+  if (!CUDAGRAPHS_MODE) {
+    // If not in cudagraphs mode, proceed with enqueueV3 as normal
     c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream(inputs[0].device().index());
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:229]: Can we take this oppurtunity to fix the default stream issue? I think we need to create a new stream for the enqueue operation

File: core/runtime/execute_engine.cpp [hunk 3]
```diff
@@ -168,27 +210,66 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
...
-    std::unique_lock<std::mutex> lock(compiled_engine->mu);
     compiled_engine->exec_ctx->enqueueV3(stream);
-    if (compiled_engine->profile_execution) {
-      LOG_INFO(std::endl << *compiled_engine->trt_engine_profiler);
-      dump_trace(compiled_engine->trt_engine_profile_path, *compiled_engine->trt_engine_profiler);
-      compiled_engine->dump_engine_layer_info();
+  } else if (need_cudagraphs_record) {
+    // If cudagraphs needs to record a graph, capture the enqueueV3 call in a graph
+
+    // Cudagraphs cannot record on the default stream, so use an alternate
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:234]: Lets reuse this for both modes

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 4]
```diff
@@ -270,3 +325,23 @@ def get_layer_info(self) -> str:
...
+
+    def cudagraphs_validate_shapes(self, inputs: Sequence[torch.Tensor]) -> bool:
+        """
+        Validates the input shapes of the forward function
+        versus the version currently active for the
+        """
+        # Representation of input shapes to a given model
+        # Shapes are concatenated as so:
+        # x: (3, 4), y: (4, 5) --> Key: (3, 4)(4, 5)
+        new_shape_key = "".join(str(tuple(t.shape)) for t in inputs)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:337]: Can we have no whitespace in the key, ensures that there is only one valid representation of a key

File: py/torch_tensorrt/runtime/cudagraphs.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import logging
+from importlib.util import find_spec
+from typing import Any
+
+import torch
+
+if find_spec("torch_tensorrt._C") is not None:
```
Comment (narendasan) [inline py/torch_tensorrt/runtime/cudagraphs.py:7]: You can use the ENABLED_FEATURES structure for this

File: py/torch_tensorrt/runtime/cudagraphs.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+logger = logging.getLogger(__name__)
+
+
+def set_cudagraphs_mode(mode: bool) -> None:
+    # Set new cudagraphs mode for Python
+    global _PY_RT_CUDAGRAPHS
+    _PY_RT_CUDAGRAPHS = mode
+
+    # Set new mode for C++
+    if find_spec("torch_tensorrt._C") is not None:
```
Comment (narendasan) [inline py/torch_tensorrt/runtime/cudagraphs.py:22]: same here

File: tests/py/dynamo/runtime/test_cudagraphs.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
+import unittest
+
+import torch
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+import torch_tensorrt
+
+from ..testing_utilities import DECIMALS_OF_AGREEMENT
+
+
```
Comment (narendasan) [inline tests/py/dynamo/runtime/test_cudagraphs.py:10]: We should only skip the CPP tests specifically right? What about python runtime tests?

File: core/runtime/execute_engine.cpp [hunk 8]
```diff
@@ -58,6 +59,29 @@ RTDevice select_rt_device(const RTDevice& engine_device, const RTDevice& curr_de
...
 }
 
+bool _cudagraphs_validate_shapes(std::vector<at::Tensor> inputs, c10::intrusive_ptr<TRTEngine> compiled_engine) {
+  // Validate whether the current input shapes to the engine
+  // invalidate the existing cudagraphs object
+
+  // Populate the shape key for the inputs
+  std::stringstream new_shape_key_ss;
+  for (auto input : inputs) {
+    new_shape_key_ss << input.sizes();
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:69]: Does this create the same shape key as python?

File: core/runtime/execute_engine.cpp (hunk 1 above)
Comment (gs-olive) [reply core/runtime/execute_engine.cpp:214]: If not in cudagraphs mode, the main downside is the requirement that buffers need to change size/be reallocated in Dynamic shape usecases.

My understanding is that these I/O buffers need to be reserved for the duration of the cudagraph being active

File: core/runtime/execute_engine.cpp (hunk 2 above)
Comment (gs-olive) [reply core/runtime/execute_engine.cpp:229]: Yes, I think in both the Python and C++ cases, we can use a non-default stream. I am not sure if we want to grab a new stream from the pool on every `forward` pass, or if we should store the stream after first inference call.

File: core/runtime/execute_engine.cpp (hunk 8 above)
Comment (gs-olive) [reply core/runtime/execute_engine.cpp:69]: I believe so - I will check the bracketing to see if `.sizes()` uses `[]` or `()`, but the scheme is the same. I will remove the whitespace in this case as well.

File: core/runtime/execute_engine.cpp (hunk 2 above)
Comment (gs-olive) [reply core/runtime/execute_engine.cpp:229]: `self.stream`
--> Synchronize at I/O boundaries
```python
stream = torch.cuda.Stream()
torch.cuda.set_stream(stream)
```

Comment (narendasan) [reply core/runtime/execute_engine.cpp:229]:
```py
torch.cuda.synchronize()
with torch.cuda.stream(s1):
    C = torch.mm(A, A)
```

**Review by peri044 — COMMENTED** [review]
LGTM

**Review by peri044 — APPROVED** [review]
LGTM

**Review by zewenli98 — APPROVED** [review]
LGTM
